File: soal_3.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

// Struktur data untuk reservasi
struct Reservasi {
    string nomor_kamar;
    string nama_tamu;
    double harga_per_malam;
    int jumlah_malam;
};

// Vector untuk menyimpan reservasi
vector<Reservasi> reservasi;

// Fungsi untuk mencatat pemesanan kamar baru (soal a)
void catat_reservasi() {
    string nomor_kamar, nama_tamu;
    double harga_per_malam;
    int jumlah_malam;

    cout << "Masukkan nomor kamar: ";
    cin >> nomor_kamar;
    cout << "Masukkan nama tamu: ";
    cin >> nama_tamu;
    cout << "Masukkan harga per malam: ";
    cin >> harga_per_malam;
    cout << "Masukkan jumlah malam: ";
    cin >> jumlah_malam;

    reservasi.push_back({nomor_kamar, nama_tamu, harga_per_malam, jumlah_malam});
    cout << "Reservasi berhasil dicatat." << endl;
}

// Fungsi untuk menampilkan seluruh reservasi (soal b)
void tampil_reservasi() {
    cout << "\nDaftar seluruh reservasi:\n";
    for (const auto& r : reservasi) {
        cout << "Nomor kamar: " << r.nomor_kamar << ", Nama tamu: " << r.nama_tamu 
             << ", Harga per malam: " << r.harga_per_malam << ", Jumlah malam: " << r.jumlah_malam << endl;
    }
}

// Fungsi untuk mencari kamar yang paling sering dipesan (soal c)
void kamar_tersering() {
    if (reservasi.empty()) {
        cout << "Tidak ada reservasi.\n";
        return;
    }

    unordered_map<string, int> freq;
    for (const auto& r : reservasi) {
        freq[r.nomor_kamar]++;
    }

    string kamar_terbanyak;
    int max_freq = 0;
    for (const auto& pair : freq) {
        if (pair.second > max_freq) {
            kamar_terbanyak = pair.first;
            max_freq = pair.second;
        }
    }

    cout << "\nKamar yang paling sering dipesan adalah kamar nomor " << kamar_terbanyak 
         << " dengan " << max_freq << " kali pemesanan.\n";
}

// Fungsi untuk menghitung total pendapatan hotel (soal d)
void total_pendapatan() {
    double total = 0;
    for (const auto& r : reservasi) {
        total += r.harga_per_malam * r.jumlah_malam;
    }
    cout << "\nTotal pendapatan hotel adalah: " << total << endl;
}

// Fungsi untuk memperbarui status ketersediaan kamar (soal e)
void update_ketersediaan_kamar(const string& nomor_kamar) {
    bool found = false;
    for (const auto& r : reservasi) {
        if (r.nomor_kamar == nomor_kamar) {
            cout << "Ketersediaan kamar nomor " << nomor_kamar << " telah diperbarui." << endl;
            found = true;
            break;
        }
    }
    if (!found) {
        cout << "Kamar nomor " << nomor_kamar << " tidak ditemukan." << endl;
    }
}

// Main program
int main() {
    int jumlah_reservasi;

    cout << "Berapa banyak reservasi yang ingin dicatat? ";
    cin >> jumlah_reservasi;

    cout << "\n-------Soal a-------\n";
    for (int i = 0; i < jumlah_reservasi; i++) {
        cout << "Reservasi ke-" << i + 1 << ":\n";
        catat_reservasi();
    }

    cout << "\n-------Soal b-------\n";
    tampil_reservasi();

    cout << "\n-------Soal c-------\n";
    kamar_tersering();

    cout << "\n-------Soal d-------\n";
    total_pendapatan();

    cout << "\n-------Soal e-------\n";
    string nomor_kamar_update;
    cout << "Masukkan nomor kamar yang ingin diupdate ketersediaannya: ";
    cin >> nomor_kamar_update;
    update_ketersediaan_kamar(nomor_kamar_update);

    return 0;
}

File: soal_3.py
# List untuk menyimpan data reservasi
reservasi = []

# Fungsi untuk mencatat pemesanan kamar baru (soal a)
def catat_reservasi():
    nomor_kamar = input("Masukkan nomor kamar: ")
    nama_tamu = input("Masukkan nama tamu: ")
    harga_per_malam = float(input("Masukkan harga per malam: "))
    jumlah_malam = int(input("Masukkan jumlah malam: "))
    
    reservasi.append({
        "nomor_kamar": nomor_kamar,
        "nama_tamu": nama_tamu,
        "harga_per_malam": harga_per_malam,
        "jumlah_malam": jumlah_malam,
    })
    print("Reservasi berhasil dicatat.")

# Fungsi untuk menampilkan seluruh reservasi (soal b)
def tampil_reservasi():
    print("\nDaftar seluruh reservasi:")
    for r in reservasi:
        print(f"Nomor kamar: {r['nomor_kamar']}, Nama tamu: {r['nama_tamu']}, Harga per malam: {r['harga_per_malam']}, Jumlah malam: {r['jumlah_malam']}")

# Fungsi untuk mencari dan menampilkan kamar yang paling sering dipesan (soal c)
def kamar_tersering():
    if not reservasi:
        print("Tidak ada reservasi.")
        return

    freq = {}
    for r in reservasi:
        if r['nomor_kamar'] in freq:
            freq[r['nomor_kamar']] += 1
        else:
            freq[r['nomor_kamar']] = 1

    kamar_terbanyak = max(freq, key=freq.get)
    print(f"Kamar yang paling sering dipesan adalah kamar nomor {kamar_terbanyak} dengan {freq[kamar_terbanyak]} kali pemesanan.")

# Fungsi untuk menghitung total pendapatan hotel (soal d)
def total_pendapatan():
    total = sum(r['harga_per_malam'] * r['jumlah_malam'] for r in reservasi)
    print(f"\nTotal pendapatan hotel adalah: {total}")

# Fungsi untuk memperbarui status ketersediaan kamar (soal e)
def update_ketersediaan_kamar(nomor_kamar):
    found = False
    for r in reservasi:
        if r['nomor_kamar'] == nomor_kamar:
            print(f"Ketersediaan kamar nomor {nomor_kamar} telah diperbarui.")
            found = True
            break
    if not found:
        print(f"Kamar nomor {nomor_kamar} tidak ditemukan.")

# Main program
if __name__ == "__main__":
    jumlah_reservasi = int(input("Berapa banyak reservasi yang ingin dicatat? "))

    print("\n-------Soal a-------")
    for i in range(jumlah_reservasi):
        print(f"\nReservasi ke-{i + 1}:")
        catat_reservasi()

    print("\n-------Soal b-------")
    tampil_reservasi()

    print("\n-------Soal c-------")
    kamar_tersering()

    print("\n-------Soal d-------")
    total_pendapatan()

    print("\n-------Soal e-------")
    nomor_kamar_update = input("Masukkan nomor kamar yang ingin diupdate ketersediaannya: ")
    update_ketersediaan_kamar(nomor_kamar_update)
